cuqi/experimental/mcmc/_langevin_algorithm.py

```diff
@@ -225,3 +249,140 @@ def _log_proposal(self, theta_star, theta_k, g_logpi_k):
+                    geometry=copied_value.prior.geometry)
+            self._smoothed_target = copied_value
+
+            # Validate the target
+            self.validate_target()
+            
+    def validate_target(self):
+        # Call ULANew target validation
+        super().validate_target()
+
+        # Additional validation for MYULA target
+        if isinstance(self.target.prior, MoreauYoshidaPrior):
+            raise ValueError(("The prior is already smoothed, apply"
+                              " ULA when using a MoreauYoshidaPrior."))
+        if not isinstance(self.target.prior, RestorationPrior):
+            raise NotImplementedError(("Using MYULA with other than"
+                                       " RestorationPrior prior"
+                                       " is not implemented yet."))
+
+    def _eval_target_grad(self, x):
+        if isinstance(self.target.prior, cuqi.implicitprior.RestorationPrior) \
+        and not np.isclose(self.smoothing_strength, self.target.prior.restoration_strength):
+            raise ValueError(f"smoothing_strength of the {self.__class__.__name__}"
+                            +"must be equal to restoration_strength of the prior " 
+                            + f"{self.target.prior.__class__.__name__}.")
+        return self._smoothed_target.gradient(x)
+
+class PnPULANew(MYULANew):
+    """Plug-and-Play Unadjusted Langevin algorithm (PnP-ULA)
+    (Laumont et al., 2022)
+
+    Samples a smoothed target distribution given its smoothed logpdf gradient based on
+    Langevin diffusion dL_t = dW_t + 1/2*Nabla target.logd(L_t)dt, where W_t is
+    a `dim`-dimensional standard Brownian motion.
+    It targets a differentiable density (partially) smoothed by a convolution
+    with Gaussian kernel with zero mean and smoothing_strength variance. The
+    smoothed target density can be made arbitrarily closed to the
+    true unsmoothed target density. 
+
+    For more details see: Laumont, R., Bortoli, V. D., Almansa, A., Delon, J.,
+    Durmus, A., & Pereyra, M. (2022). Bayesian imaging using plug & play priors:
+    when Langevin meets Tweedie. SIAM Journal on Imaging Sciences, 15(2), 701-737.
+
+    Parameters
+    ----------
+
+    target : `cuqi.distribution.Distribution`
+        The target distribution to sample. The target distribution result from
+        a differentiable likelihood and prior of type RestorationPrior.
+    
+    initial_point : ndarray
+        Initial parameters. *Optional*
+
+    scale : float
+        The Langevin diffusion discretization time step (In practice, a scale of
+        1/L, where L is the Lipschitz of the gradient of the log target density
+        is recommended but not guaranteed to be the optimal choice).
+        
+    smoothing_strength : float
+        This parameter controls the smoothing strength of PnP-ULA. smoothing_strength
+        must be equal to restoration_strength of the RestoratioPrior.
```

Comment:
Same comment as previously.

Reply:
Thank you for your comment Jakob! That is a good point. We updated the constructor of RestorationPrior and removed passing restoration_strength to it as we discussed in the meeting from 08/10/2024.
Now, RestorationPrior does not contain this attribute anymore and there is no need to enforce this equality.